okhttp-logging-interceptor/src/test/java/com/squareup/okhttp/logging/HttpLoggingInterceptorTest.java

```diff
@@ -66,218 +83,406 @@
-    assertEquals("Connection: Keep-Alive", logs.get(2));
-    assertEquals("Accept-Encoding: gzip", logs.get(3));
-    assertTrue(Pattern.matches("User-Agent: okhttp/.+", logs.get(4)));
-    assertEquals("--> END GET", logs.get(5));
-    assertTrue(Pattern.matches("<-- HTTP/1\\.1 200 OK \\(\\d+ms\\)", logs.get(6)));
-    assertEquals("Transfer-encoding: chunked", logs.get(7));
-    assertEquals("Content-Type: text/plain; charset=utf-8", logs.get(8));
-    assertEquals("OkHttp-Selected-Protocol: http/1.1", logs.get(9));
-    assertTrue(Pattern.matches("OkHttp-Sent-Millis: \\d+", logs.get(10)));
-    assertTrue(Pattern.matches("OkHttp-Received-Millis: \\d+", logs.get(11)));
-    assertEquals("", logs.get(12));
-    assertEquals("Hello!", logs.get(13));
-    assertEquals("<-- END HTTP (6-byte body)", logs.get(14));
+    assertLogs(applicationLogs)
+        .hasMessageEqual("--> GET / HTTP/1.1")
+        .hasMessageEqual("--> END GET")
+        .hasMessageMatch("<-- HTTP/1\\.1 200 OK \\(\\d+ms\\)")
+        .hasMessageEqual("Transfer-encoding: chunked")
+        .hasMessageEqual("Content-Type: text/plain; charset=utf-8")
+        .hasMessageEqual("OkHttp-Selected-Protocol: http/1.1")
+        .hasMessageMatch("OkHttp-Sent-Millis: \\d+")
+        .hasMessageMatch("OkHttp-Received-Millis: \\d+")
+        .hasMessageEqual("")
+        .hasMessageEqual("Hello!")
+        .hasMessageEqual("<-- END HTTP (6-byte body)")
+        .hasNoMoreMessages();
+
+    assertLogs(networkLogs)
+        .hasMessageEqual("--> GET / HTTP/1.1")
+        .hasMessageEqual("Host: " + host)
+        .hasMessageEqual("Connection: Keep-Alive")
+        .hasMessageEqual("Accept-Encoding: gzip")
+        .hasMessageMatch("User-Agent: okhttp/.+")
+        .hasMessageEqual("--> END GET")
+        .hasMessageMatch("<-- HTTP/1\\.1 200 OK \\(\\d+ms\\)")
+        .hasMessageEqual("Transfer-encoding: chunked")
+        .hasMessageEqual("Content-Type: text/plain; charset=utf-8")
+        .hasMessageEqual("OkHttp-Selected-Protocol: http/1.1")
+        .hasMessageMatch("OkHttp-Sent-Millis: \\d+")
+        .hasMessageMatch("OkHttp-Received-Millis: \\d+")
+        .hasMessageEqual("")
+        .hasMessageEqual("Hello!")
+        .hasMessageEqual("<-- END HTTP (6-byte body)")
+        .hasNoMoreMessages();
   }
 
   private Request.Builder request() {
     return new Request.Builder().url(server.url("/"));
   }
+
+  private static LogAssert assertLogs(List<String> logs) {
+    return new LogAssert(logs);
+  }
+
+  private static class LogAssert {
+    private final List<String> logs;
+    private int index;
+
+    LogAssert(List<String> logs) {
+      this.logs = logs;
+    }
+
+    LogAssert hasMessageEqual(String expected) {
+      assertTrue("No more messages found", index < logs.size());
+      String actual = logs.get(index++);
+      assertEquals(expected, actual);
+      return this;
+    }
+
+    LogAssert hasMessageMatch(String pattern) {
```

Comment:
> Don't love that the headers & order of headers is now something that can break this test

Hmm..  I can change the tests to call the `Interceptor` API directly instead of doing these integration tests.